### ISSUE TITLE
Unable to communicate between supernode and edge

### ISSUE THREAD

**Issue by Alsan** [ntop/n2n#750]
I'm trying to setup a private vpn with n2n, but can't make it work. here's my test environment:

## host1 (both supernode and edge1)

```bash
./supernode -l 5000 -f -vvv
sudo ./edge -a ${SUBNET}.1 -c g1 -k test -l ${SERVER_IP}:5000 -f -vvv
```

## host2 (edge2)

```bash
sudo ./edge -a ${SUBNET}.11 -c g1 -k test -l ${SUPERVER_IP}:5000 -f -vvv

## test vpn
ping ${SUBNET}.1
ssh ${SUBNET}.1
```
all the test are failed.

**Comment by Logan007** [comment]
I assume you use 2.9 from _dev_ branch. The supernode needs `-p 5000` for the port, not `-l ...`. `-l` usage has been changed and now is used to provide additional supernodes to the supernode (see Federation](https://github.com/ntop/n2n/blob/dev/doc/Federation.md)) or to the edge (as you already did).

Do the logs show anything specific?

**Comment by Alsan** [comment]
ok, I'v changed to `supernode -p 5000 -f -vvv`, then do a ping from edge2 to edge1, and I can see the log for edge2 is running in per second interval, and with a little delay, the log for supernode is running, similar interval with edge2, the log for edge1 is never run, and of cause, the ping result from edge2 is 100% packet loss.

I do the test again with switched side, from edge1 to edge2, same result as expected.

**Comment by Logan007** [comment]
> the log for edge1 is never run,

Does edge1 connect to the supernode, i.e. does it show a line `[OK] Edge Peer <<< ================ >>> Super Node` at the beginning of the log?

As edge1 and supernode reside on the same host, this is really strange. If you use `localhost` or `127.0.0.1` along with `-l`, you could try to use edge1's host IP address or vice versa.

Are supernode and edge all compiled from current _dev_ (exact same version number)?

> I do the test again with switched side, from edge1 to edge2, same result as expected.

Does that mean that edge2 does not show any log activities? Or edge1?

**Comment by Alsan** [comment]
> Does edge1 connect to the supernode, i.e. does it show a line [OK] Edge Peer <<< ================ >>> Super Node at the beginning of the log?

Yes, both edge 1 and 2 is connected to supernode (has the indicator you mentioned).

> As edge1 and supernode reside on the same host, this is really strange. If you use localhost or 127.0.0.1 along with -l, you could try to use edge1's host IP address or vice versa.

I've seen some materials in the web described the usage of this scenario, and I think this is best for testing before actual setup of the network.

> Are supernode and edge all compiled from current dev (exact same version number)?

Yes, of cause.

> Does that mean that edge2 does not show any log activities? Or edge1?

Yes, it doesn't show any activities from the others (another edge) except the status report from it's own. Here's a part of the log from edge being pinged (the receiver):

```log
20/Aug/2021 05:04:08 [n2n.c:288] supernode2sock successfully resolves supernode IPv4 address for 144.34.161.226
20/Aug/2021 05:04:08 [edge_utils.c:1492] Registering with supernode [144.34.161.226:5000][number of supernodes 1][attempts left 2]
20/Aug/2021 05:04:08 [edge_utils.c:1173] send REGISTER_SUPER to 144.34.161.226:5000
20/Aug/2021 05:04:08 [edge_utils.c:957] sendto_fd sent=79 to
20/Aug/2021 05:04:08 [edge_utils.c:525] Registering with multicast group 224.0.0.68:1968
20/Aug/2021 05:04:08 [edge_utils.c:1317] Send REGISTER to 224.0.0.68:1968
20/Aug/2021 05:04:08 [edge_utils.c:957] sendto_fd sent=61 to
20/Aug/2021 05:04:08 [n2n.c:607] Purging old registrations
20/Aug/2021 05:04:08 [n2n.c:612] Remove 0 registrations
20/Aug/2021 05:04:08 [edge_utils.c:2313] ### Rx N2N UDP (61) from 144.34.161.226:41881
20/Aug/2021 05:04:08 [edge_utils.c:2448] Skipping REGISTER from self
20/Aug/2021 05:04:08 [edge_utils.c:2313] ### Rx N2N UDP (58) from 144.34.161.226:5000
20/Aug/2021 05:04:08 [edge_utils.c:2555] Rx REGISTER_SUPER_ACK from MAC 12:39:15:CE:2D:EB [144.34.161.226:5000] (external 144.34.161.226:41881). Attempts 2
```

Form the log, it's seems it can talk to the supernode, but not from the another edge.

The following log is the sender edge during ping:

```log
20/Aug/2021 17:11:42 [edge_utils.c:2232] ### Rx TAP packet (  42) for FF:FF:FF:FF:FF:FF
20/Aug/2021 17:11:42 [transform_aes.c:82] transop_encode_aes 42 bytes plaintext
20/Aug/2021 17:11:42 [edge_utils.c:2182] Encode 96 B PACKET [42 B data, 54 B overhead] transform 3
20/Aug/2021 17:11:42 [edge_utils.c:1981] Multicast or broadcast destination peer, using supernode
20/Aug/2021 17:11:42 [edge_utils.c:2041] Tx PACKET to 144.34.161.226:5000 (dest=FF:FF:FF:FF:FF:FF) [96 B]
20/Aug/2021 17:11:42 [edge_utils.c:957] sendto_fd sent=96 to
20/Aug/2021 17:11:43 [edge_utils.c:2232] ### Rx TAP packet (  42) for FF:FF:FF:FF:FF:FF
20/Aug/2021 17:11:43 [transform_aes.c:82] transop_encode_aes 42 bytes plaintext
20/Aug/2021 17:11:43 [edge_utils.c:2182] Encode 96 B PACKET [42 B data, 54 B overhead] transform 3
20/Aug/2021 17:11:43 [edge_utils.c:1981] Multicast or broadcast destination peer, using supernode
20/Aug/2021 17:11:43 [edge_utils.c:2041] Tx PACKET to 144.34.161.226:5000 (dest=FF:FF:FF:FF:FF:FF) [96 B]
20/Aug/2021 17:11:43 [edge_utils.c:957] sendto_fd sent=96 to
20/Aug/2021 17:11:44 [edge_utils.c:2232] ### Rx TAP packet (  42) for FF:FF:FF:FF:FF:FF
20/Aug/2021 17:11:44 [transform_aes.c:82] transop_encode_aes 42 bytes plaintext
20/Aug/2021 17:11:44 [edge_utils.c:2182] Encode 96 B PACKET [42 B data, 54 B overhead] transform 3
20/Aug/2021 17:11:44 [edge_utils.c:1981] Multicast or broadcast destination peer, using supernode
20/Aug/2021 17:11:44 [edge_utils.c:2041] Tx PACKET to 144.34.161.226:5000 (dest=FF:FF:FF:FF:FF:FF) [96 B]
20/Aug/2021 17:11:44 [edge_utils.c:957] sendto_fd sent=96 to
```

**Comment by Logan007** [comment]
This is really strange... 

Just for testing, you could try to force the edges to communicate via the supernode. Just use `-S` (UDP) or `-S2` (TCP) at both edges.

I wonder why the edge sends out bropadcasts (destination MAC address `FF:FF:FF:FF:FF:FF`). Those might be ARP packets. Could you please check your `arp` command output for the respectively other edge?

Also, I would be interested in the supernodes log output while sending the pings.

And finally, could you have a look at the supernodes [management port output](https://github.com/ntop/n2n/blob/dev/doc/Faq.md#can-i-get-a-list-of-connected-edge-nodes-and-their-community-and-source-ip-address-from-the-supernode)? I wonder if both edges are properly registered within the same community.

**Comment by Alsan** [comment]
I just do the same test this morning in my home, and all the ping and ssh are worked just fine. So, the real problem is in my office network. I'm off these two days (sat and sun), and will post the log while I'm back to work.

**Comment by Alsan** [comment]
Just test today in my office, with -S2 in both edge, all the tests worked. That said, the UDP is blocked by someone in the middle. I just wonder I have test the UDP port connectivity with nc (netcat) to verify does n2n working, before my post, and nc is working for the same port, but why not n2n, is there any signature being captured?

Thanks a lot for the help.

**Comment by Logan007** [comment]
> in my office [...] UDP is blocked by someone in the middle

Yes, maybe it is blocked by some switch or so.

On the other hand, blocking does not necessarily have to happen in the middle, could be local computer firewall policy on external network interface (not: localhost) as well. If this is the case, you could assign a fixed port, e.g. `-p 2000`, and make the firewall exempt this specific port from its regime to allow full free UDP access.

Also, you could try a lower MTU `-M ...` with a value lower than 1290. But this is unlikely to help as 1290 already is kind of low and ping with its short packets should work anyway.

> any signature being captured?

I have not heard of any n2n specific filters yet. But you can check and enable [header encrpytion](https://github.com/ntop/n2n/blob/dev/doc/Crypto.md#header) (`-H` at _both_ edges and community name inside the `-c <file>`-provided file at supernode, restart the supernode).

**Comment by Alsan** [comment]
I just did another test with -S1 in both edges, and all works as expected. That said, there's nothing related to UDP but method of finding peer, or how it communicating with the peer.

I might try to lean about the source code on finding what's happening, and how to do with going into peer-to-peer mode.

**Comment by Logan007** [comment]
> I just did another test with -S1 in both edges, and all works as expected.

That's good news on the one hand, a bit strange on the other hand because the setup should fall back to supernode forwarding automatically if p2p not working...

Please make sure that there is no local firewall rule preventing unwanted incoming or outgoing UDP traffic at one of the computers, such as a non-symmetric firewall or NAT.

> I might try to lean about the source code on finding what's happening, and how to do with going into peer-to-peer mode.

I'd be very interested in your findings. `process_udp()` at `src/edge_utils.c:2273` handles incoming packets and would be a good starting point to add some `printf`s or debugger break points – depending on your preference. Similarly, `src/sn_utils.c:1630` for the supernode.

**Comment by Alsan** [comment]
> Please make sure that there is no local firewall rule preventing unwanted incoming or outgoing UDP traffic at one of the computers, such as a non-symmetric firewall or NAT.

I can confirm that there's no any local firewall between the testing computers, and that's why I'm setting up the supernode and one of the edge in some box for the testing environment.

> I'd be very interested in your findings. process_udp() at src/edge_utils.c:2273 handles incoming packets and would be a good starting point to add some printfs or debugger break points – depending on your preference. Similarly, src/sn_utils.c:1630 for the supernode.

Thanks for the tips. I'll post my finding once I'm fingering out what is happening.

**Comment by Logan007** [comment]
>  that's why I'm setting up the supernode and one of the edge in some box for the testing environment.

Please remember to not use `-l 127.0.0.1:port` or `-l localhost:port` but `-l <interface IP address>:port` at that edge then.

**Comment by Logan007** [comment]
Is this still an issue?

**Comment by Alsan** [comment]
Yes, the problem still exist. It seems that multiple factors cause the issue. The windows implementation always get an empty frame when keep alive signal, and the linux implementation can't get the real external ip correctly, so that the p2p communication not work, but fallback to server side proxy can work correctlly.

I'm guessing that this is caused by multiple layer of NAT -- the public ip can only valid in that NAT layer.

**Comment by Logan007** [comment]
> always get an empty frame when keep alive signal

What exactly is the "empty frame"? To what packet type are you referring with the "keep alive signal", REGISTER or REGISTER_SUPER?

> Windows
> Linux

There were reports about incompatible builds solved by #830. Did you try newest version?

> can't get the real external ip correctly
> multiple layer of NAT

I am not sure about your network setup and where exactly your supernode and edges are placed in it. But usually, multiple NAT has not been an issue so far. Do you want to describe it in a bit more detail?

**Comment by Alsan** [comment]
> What exactly is the "empty frame"? To what packet type are you referring with the "keep alive signal", REGISTER or REGISTER_SUPER?

I'm sorry that my server has been disposed due to resources outage with other services running, have to do the test and get the log latter.

> There were reports about incompatible builds solved by #830. Did you try newest version?

I didn't update the code base for a while, I'll try it later.

> I am not sure about your network setup and where exactly your supernode and edges are placed in it. But usually, multiple NAT has not been an issue so far. Do you want to describe it in a bit more detail?

The network architecture between the suppernode and edges are quite simple and classic, just direct dail from the edges (one linux edge node and one windows edge node) to the supernode. But the actual network topology is quite complicated and out of my control.

First of all, I'm living in China, the two edages are in China, but the supernode is setup at USA, and the multiple layers of the NAT is caused by the GFW.

The both edges are using dialup connection, and the routes to the supernode are crossing at least 17 routers and at least half of them are in China. I'm guessing that some of these routers in the China side will form a layer of NAT for each cell (district, city, state, country).

**Comment by kamikazebr** [comment]
I looks like have same issue.

I'm CentOS 7 in the cloud and my laptop Ubuntu 20.04

Supernova and Edge in Centos

Edge in Ubuntu

All installed by yum and apt

In edge Ubuntu i get

![image](https://user-images.githubusercontent.com/691510/142575551-7924972f-36da-4b4c-96cc-5dc1dce36af9.png)

With verbose (-v)

![image](https://user-images.githubusercontent.com/691510/142575622-8c2d2dee-b3ff-49f9-be8c-dc4516a44eba.png)

**Comment by hamishcoleman** [comment]
@kamikazebr , What version of the n2n edge and supernode software are you using?  Did you download the dpkg and rpm from somewhere, or are you using the one from the OS repositories (for example, the one in the Ubuntu repository is years out of date)

What options or config files are you using to start your edge daemons?

Can you describe exactly what problem you are experiencing?

**Comment by kamikazebr** [comment]
> @kamikazebr , What version of the n2n edge and supernode software are you using? Did you download the dpkg and rpm from somewhere, or are you using the one from the OS repositories (for example, the one in the Ubuntu repository is years out of date)
> 
> What options or config files are you using to start your edge daemons?
> 
> Can you describe exactly what problem you are experiencing?

@hamishcoleman 
Sorry for mega delay, but i just move on and dont remember exactly the config.

**Comment by hamishcoleman** [comment]
@kamikazebr From the messages and code line numbers in your screenshots of the log output, it looks like you are using a version 1 of n2n - the project has long ago upgraded to version 3 and cannot really support version 1.

Additionally, the major versions are not compatible with each other, so that might also have been the issue you experienced (I think that the Ubuntu and Centos repos both have different major versions of n2n packaged)

**Comment by hamishcoleman** [comment]
@kamikazebr From the messages and code line numbers in your screenshots of the log output, it looks like you are using a version 1 of n2n - the project has long ago upgraded to version 3 and cannot really support version 1.

Additionally, the major versions are not compatible with each other, so that might also have been the issue you experienced (I think that the Ubuntu and Centos repos both have different major versions of n2n packaged)

**Comment by hamishcoleman** [comment]
@kamikazebr From the messages and code line numbers in your screenshots of the log output, it looks like you are using a version 1 of n2n - the project has long ago upgraded to version 3 and cannot really support version 1.

Additionally, the major versions are not compatible with each other, so that might also have been the issue you experienced (I think that the Ubuntu and Centos repos both have different major versions of n2n packaged)
